Login/Payment protocol for DSPs (#331)
applicaster/developer.applicaster.com: book/Zapp-Pipes/protocols/login_and_payment_protocol.md

# Login and Payment Protocol

> This document outlines the requirements for mapping the data to determine whether login and/or payment are required to gain access to a video item.

Reviewer: Is it limited to video items? What about articles or other playable items like podcasts?

Author: You are right, I will change it. Not limited to videos only.

## The "Free" flag
Previously, the flag that was used to achieve this was the "free" flag in the extensions section of the item. However, there are a few drawbacks for using the "free" flag:

> 1. The flag doesn't distinguish between whether an item requires login or payment to get access to, and might create some confusion. 

Reviewer: ```suggestion
1. The flag doesn't distinguish between whether access to an item requires login or payment, and this might create some confusion. 
```

2. If the item requires purchase, the "free" flag doesn't specify what entitlements are needed in order to unlock the item.

The suggested protocol is meant to be used by DSPs, who want to provide this information to the app, and can be used by Login / Payment plugins to interpret the action that is needed for the item to be unlocked. 

**Important Note**: if you are using Applicaster2 as your datasource provider, there is no way to add the new extensions to an item, and therefore the "free" flag is the only way to configure an item as "locked".

## "requires_authentication" and "ds_product_ids"

The protocol is based on two fields that are added to the "extensions" section of the item:

* **requires_authentication**: a boolean flag to determine whether login is required to unlock the item. 
This field should be of type "boolean" (true/false) and *not* a "string".
If the flag is not specified, the default value is `false`, which means the item does not require authentication.

* **ds_product_ids**: a list of entitlements needed to unlock the item. 

> It is up to the Login / Payment provider to decide if *all* entitlements are required, or a single entitlement is sufficient. 

Reviewer: Oh really? I didn't know that. I thought one of the entitlements would suffice. How would it work if two entitlements are required?

Author: Maybe it's not a valid case, and the only valid case is a "OR" between the entitlements. I didn't want to limit that, because technically it's not the DSP's concern how the entitlements data is used. I would change the text to say something "typically a single entitlement fulfilled should give access to the item"

Author: xI am not referring to this anymore in the document

> If the field is not specified, this means that the item does not require purchase.

Reviewer: ```suggestion
If the field is not provided or is left blank, this means that the item does not require purchase.
```

Unless specified is understood to mean that for developers. As a product person, I would have interpreted it as populated.

### Example:
Here is an example of how the field will be configured on a video item:

```
{
  "type": {
    "value": "video"
  },
  "id": "5718802496001",
  "title": "Video that requires login and purchase",
  "published": "2018-01-24T20:48:01+00:00",
  "extensions": {
		“requires_authentication”: true,
		“ds_product_ids”: [“entitlement1”,“entitlement2”,"entitlement3”]
    }
}
```
In the example above, the video item requires both login and payment. 

> The list of entitlements that are attached to the item, are listed as "entitlement1",  "entitlement2" and "entitlement3".

Reviewer: ```suggestion
The list of entitlements that are attached to the item, are listed as "entitlement1",  "entitlement2" and "entitlement3", but typically they are some sort of code/ID. When possible, we recommend aligning this code/ID to what the store and plugin provider use in order to eliminate the need for mapping across systems.
```

# Login and Payment Protocol

> This document outlines the requirements for mapping the data to determine whether login and/or payment are required to gain access to a video item.

Reviewer: ```suggestion
This document outlines the requirements for mapping the data which determines whether login and/or payment are required to gain access to a video item.
```

Is it just for video items? Would it work for audio playables or articles or photo galleries?

## The "Free" flag

> Previously, the flag that was used to achieve this was the "free" flag in the extensions section of the item. However, there are a few drawbacks for using the "free" flag:

Reviewer: ```suggestion
Previously, the flag that was used to achieve this was the "free" flag in the extensions section of the item. However, there are a few drawbacks to using the "free" flag:
```

1. The flag doesn't distinguish between whether access to an item requires login or payment, and this might create some confusion. 
2. If the item requires purchase, the "free" flag doesn't specify what entitlements are needed in order to unlock the item.

> The suggested protocol is meant to be used by DSPs, who want to provide this information to the app, and can be used by Login / Payment plugins to interpret the action that is needed for the item to be unlocked. 

Reviewer: ```suggestion
The protocol defined below is meant to be used by DSPs who want to provide this information to the app, and can be used by Login / Payment plugins to interpret the action that is needed for the item to be unlocked. 
```
Perhaps described is more accurate than defined. I feel that if we want it to become standard, suggested is not a strong enough word. Suggested is like recommended, but even gentler.

**Important Note**: if you are using Applicaster2 as your datasource provider, there is no way to add the new extensions to an item, and therefore the "free" flag is the only way to configure an item as "locked".

## "requires_authentication" and "ds_product_ids"

The protocol is based on two fields that are added to the "extensions" section of the item:

* **requires_authentication**: a boolean flag to determine whether login is required to unlock the item. 
This field should be of type "boolean" (true/false) and *not* a "string".
If the flag is not specified, the default value is `false`, which means the item does not require authentication.

* **ds_product_ids**: a list of entitlements needed to unlock the item.

> Typically they are some sort of code/ID. 

Reviewer: ```suggestion
Typically an entitlement is some sort of code/ID. 
```

When possible, we recommend aligning this code/ID to what the store and plugin provider use in order to eliminate the need for mapping across systems.
If the field is not provided or is left blank, this means that the item does not require purchase.

> 

Reviewer: You might want to mention that more than one entitlement can be provided and how do so (e.g. comma separated).

Also, we could mention here that multiple entitlements are interpreted to mean that _any_ of those entitlements provide authorization, not that _all_ are required to provide authorization.

Author: Done! Thanks

### Example:
Here is an example of how the field will be configured on a video item:

```
{
  "type": {
    "value": "video"
  },
  "id": "5718802496001",
  "title": "Video that requires login and purchase",
  "published": "2018-01-24T20:48:01+00:00",
  "extensions": {
		“requires_authentication”: true,
		“ds_product_ids”: [“entitlement1”,“entitlement2”,"entitlement3”]
    }
}
```
In the example above, the video item requires both login and payment. 

> The list of entitlements that are attached to the item, are listed as "entitlement1",  "entitlement2" and "entitlement3", 

Reviewer: ```suggestion
The list of entitlements that are attached to the item, are listed as "entitlement1",  "entitlement2", and "entitlement3".
```
